src/Engine/Combat/CombatContext.java

```diff
@@ -118,6 +124,32 @@ else if( calcType == CalcType.OPTIMISTIC )
       calcType = CalcType.PESSIMISTIC;
   }
 
+  public static void setTowerCounts(GameMap map, UnitContext uc)
+  {
+    // Count number of towers, and apply new modifiers
+    final int minX = 0;
+    final int minY = 0;
+    final int maxX = map.mapWidth  - 1;
+    final int maxY = map.mapHeight - 1;
+
+    uc.towerCountDoR = 0;
+    uc.towerCountDS  = 0;
+    for( int y = minY; y <= maxY; y++ ) // Top to bottom, left to right
+    {
+      for( int x = minX; x <= maxX; x++ )
+      {
+        MapLocation loc = map.getLocation(x, y);
+        Commander owner = loc.getOwner();
+        if( null == owner || uc.CO.army != owner.army )
+          continue;
+        if( loc.getEnvironment().terrainType.equals(TerrainType.DOR_TOWER) )
+          uc.towerCountDoR += 1;
+        if( loc.getEnvironment().terrainType.equals(TerrainType.DS_TOWER) )
+          uc.towerCountDS += 1;
+      }
+    }
```

Comment:
Seems like tower counts would be a useful property to just have on the Commander. This would also let us recount much less frequently.

Reply:
On the one hand, yes.
On the other hand, caching that value lead to weirdness in the past.
On the third hand, this collects the comm count *according to a specific observer*, so the comm count isn't leaked by combat previews.